File: Tests/server/executors/AssetsExecutor.cpp
#include "AssetsExecutor.h"

#include <format>
#include <numeric>

static std::string customFunction(const std::vector<std::string>& args)
{
	return std::format("Data: {} {} {}", args[0], args[1], args[2]);
}

void AssetsExecutor::doGet(framework::HTTPRequest& request, framework::HTTPResponse& response)
{
	for (const auto& [key, value] : request.getKeyValueParameters())
	{
		std::cout << key << ' ' << value << std::endl;
	}

	request.sendDynamicFile
	(
		std::format("{}.wfdp", request.getJSON().getString("fileName")),
		response,
		request.getKeyValueParameters()
	);
}

void AssetsExecutor::doPost(framework::HTTPRequest& request, framework::HTTPResponse& response)
{
	request.registerDynamicFunction("customFunction", customFunction);
}

void AssetsExecutor::doDelete(framework::HTTPRequest& request, framework::HTTPResponse& response)
{
	request.unregisterDynamicFunction("customFunction");
}

DECLARE_EXECUTOR(AssetsExecutor)

File: Tests/executors/CSharp/CoreExecutors/AssetsExecutor.cs
using System.Text.Json.Nodes;

using CoreExecutors;
using Framework;

public class AssetsExecutor : StatelessExecutor
{
	public override void Init(ExecutorSettings settings)
	{
		settings.RegisterDynamicFunction<CustomFunction>("customFunction");
	}

	public override void DoGet(HttpRequest request, HttpResponse response)
	{
		IDictionary<string, string> queryParameters = request.GetQueryParameters();
		string fileName = request.GetJson()["fileName"].GetString()!;
		JsonObject arguments = [];

		if (queryParameters.ContainsKey("data"))
		{
			arguments["@print"] = new JsonObject()
			{
				["data"] = queryParameters["data"]
			};
		}
		else if (queryParameters.ContainsKey("first"))
		{
			arguments["@customFunction"] = new JsonObject()
			{
				["data"] = new JsonArray
				(
					int.Parse(queryParameters["first"]),
					int.Parse(queryParameters["second"]),
					int.Parse(queryParameters["third"])
				)
			};
		}

		request.SendDynamicFile($"{fileName}.wfdp", response, arguments);
	}

	public override void DoDelete(HttpRequest request, HttpResponse response)
	{
		request.UnregisterDynamicFunction("customFunction");
	}
}
